utilities/blob_db/blob_db.cc

```diff
@@ -17,194 +20,139 @@
-
-  using rocksdb::StackableDB::Get;
-  Status Get(const ReadOptions& options, const Slice& key,
-             std::string* value) override;
-
-  Status Open();
-
-  explicit BlobDB(DB* db);
-
- private:
-  std::string dbname_;
-  ImmutableCFOptions ioptions_;
-  InstrumentedMutex mutex_;
-  std::unique_ptr<RandomAccessFileReader> file_reader_;
-  std::unique_ptr<WritableFileWriter> file_writer_;
-  size_t writer_offset_;
-  size_t next_sync_offset_;
-
-  static const std::string kFileName;
-  static const size_t kBlockHeaderSize;
-  static const size_t kBytesPerSync;
-};
-
-Status NewBlobDB(Options options, std::string dbname, DB** blob_db) {
-  DB* db;
-  Status s = DB::Open(options, dbname, &db);
-  if (!s.ok()) {
-    return s;
-  }
-  BlobDB* bdb = new BlobDB(db);
-  s = bdb->Open();
-  if (!s.ok()) {
-    delete bdb;
-  }
-  *blob_db = bdb;
-  return s;
-}
+namespace blobstorage {
+port::Mutex listener_mutex;
+typedef std::shared_ptr<BlobDBFlushBeginListener> FlushBegin_t;
+typedef std::shared_ptr<BlobReconcileWalFilter> ReconcileWal_t;
+typedef std::shared_ptr<EvictAllVersionsFilterFactory> EvictAllVersions_t;
 
-const std::string BlobDB::kFileName = "blob_log";
-const size_t BlobDB::kBlockHeaderSize = 8;
-const size_t BlobDB::kBytesPerSync = 1024 * 1024 * 128;
-
-BlobDB::BlobDB(DB* db)
-    : StackableDB(db),
-      ioptions_(db->GetOptions()),
-      writer_offset_(0),
-      next_sync_offset_(kBytesPerSync) {}
-
-Status BlobDB::Open() {
-  unique_ptr<WritableFile> wfile;
-  EnvOptions env_options(db_->GetOptions());
-  Status s = ioptions_.env->NewWritableFile(db_->GetName() + "/" + kFileName,
-                                            &wfile, env_options);
-  if (!s.ok()) {
-    return s;
-  }
-  file_writer_.reset(new WritableFileWriter(std::move(wfile), env_options));
+// to ensure the lifetime of the listeners
+std::vector<FlushBegin_t> all_flush_begins;
+std::vector<ReconcileWal_t> all_wal_filters;
 
-  // Write version
-  std::string version;
-  PutFixed64(&version, 0);
-  s = file_writer_->Append(Slice(version));
-  if (!s.ok()) {
-    return s;
-  }
-  writer_offset_ += version.size();
+Status BlobDB::OpenAndLoad(const Options& options,
+                           const BlobDBOptions& bdb_options,
+                           const std::string& dbname, BlobDB** blob_db,
+                           Options* changed_options) {
+  *changed_options = options;
+  *blob_db = nullptr;
 
-  std::unique_ptr<RandomAccessFile> rfile;
-  s = ioptions_.env->NewRandomAccessFile(db_->GetName() + "/" + kFileName,
-                                         &rfile, env_options);
-  if (!s.ok()) {
-    return s;
+  FlushBegin_t fblistener = std::make_shared<BlobDBFlushBeginListener>();
+  ReconcileWal_t rw_filter = std::make_shared<BlobReconcileWalFilter>();
+  EvictAllVersions_t comp_filt_factory =
+      std::make_shared<EvictAllVersionsFilterFactory>();
+
+  {
+    MutexLock l(&listener_mutex);
+    all_flush_begins.push_back(fblistener);
```

Comment:
If I understand you correctly, you are asking for move semantics. I am using fblistener and rw_filter and move will null out the shared pointers.